Volume Rendering datasets in spherical coordinates

chrishavlin opened the pull request:
So this is a verion of #64 that seems to work pretty well with a few caveats. 

This version does 2 things differently: 

1. It calculates the cartesian bounding boxes of every block initially (using code from https://github.com/yt-project/yt/pull/4847), and passes that info down to the shaders as additional vertex attributes.
2. And then those cartesian bounding boxes are used in the ray tracing shader for calculating ray intersections. At each sample point along the ray, the spherical coordinates are calculated and used to sample the texture if it falls in the actual spherical element.

The main caveat is that you need a fairly large number of sample points for the ray tracing since intersection with the cartesian bounding box does not guarantee intersection with the enclosed spherical element. 

EDIT: cartsian bbox calculation now works with big elements (via a recursive division when its needed). ~~Also, at least one thing that should be fixed before this goes in is that the cartesian bounding box calculation was designed for fairly small spherical elements and breaks down when blocks span whole quadrants of the spherical coordinate system. e.g., the new example script uses `load_uniform_grid` with `nproc=128`, if `nproc` is smaller (less than 8 or so), the blocks get too big and the bounding box calculation breaks down and results start to get wonky (or nonexistant). So, should fix that...~~

Finally -- I decided to go with a new PR rather than modifying #64 because I'm abandoning the complicated intersection checks with the underlying spherical element bounds. While I think that approach **could** work (and could be added on here...),  I think it's worth getting the simpler version here merged first and wanted to preserve #64 in its current state to make that attempt easier in the future.

chrishavlin commented on yt_idv/scene_data/block_collection.py:
```diff
@@ -5,6 +5,7 @@
 from yt.data_objects.data_containers import YTDataContainer
 
 from yt_idv.opengl_support import Texture3D, VertexArray, VertexAttribute
+from yt_idv.scene_data._geometry_utils import phi_normal_planes
```
note to self: can remove the phi-normal plane calculations for this PR -- the code will exist in #64 should we want to add it back in the future...

chrishavlin commented on yt_idv/scene_data/block_collection.py:
```diff
@@ -74,30 +76,125 @@ def add_data(self, field, no_ghost=False):
+            )
+
+            axis_id = self.data_source.ds.coordinates.axis_id
+            phi_plane_le = phi_normal_planes(le, axis_id, cast_type="f4")
+            phi_plane_re = phi_normal_planes(re, axis_id, cast_type="f4")
+            self.vertex_array.attributes.append(
+                VertexAttribute(name="phi_plane_le", data=phi_plane_le)
+            )
+            self.vertex_array.attributes.append(
+                VertexAttribute(name="phi_plane_re", data=phi_plane_re)
```
one more note -- will rm these, they're not being used by the shader anymore.

matthewturk commented:
screenshots???

chrishavlin commented:
running `python amr_spherical_volume_rendering.py -f stream,tube`: 

https://github.com/user-attachments/assets/735614cb-866c-4fea-b3cc-92a88f8f261c

switching to the constant shader at the end of the video shows the actual blocks that come from the cartesian bounding boxes.

matthewturk commented:
oh my word, @chrishavlin .  oh my word.
![poet-jodie-foster](https://github.com/user-attachments/assets/28fa4f27-7249-49d2-98f8-655cceb10fa9)

ivanov commented:
@matthewturk sent me this message:
> this is **expletive** amazing and I love it and I need to share it with someone.  ray casting spherical domains is SO HARD.

good job, @chrishavlin , all the :cake: for you!

chrishavlin commented:
and another, `python amr_spherical_volume_rendering.py -f stream,r_rev -d quadrant_shell`


https://github.com/user-attachments/assets/c92da8ad-645a-46f8-be5e-53cad08a5658

the field here is "reversed radius" (`max r - r`), which makes it a little easier to see the 3d-ness.

chrishavlin commented on yt_idv/shaders/grid_expand.geom.glsl:
```diff
@@ -36,22 +40,35 @@ uniform vec3 arrangement[8] = vec3[](
+    vec3 width, le;
     vec4 newPos;
 
+    if (is_spherical){
```
one more reminder for myself: go look at the pre-processor business again (e.g., https://github.com/yt-project/yt_idv/pull/89), might not need uniforms here.

chrishavlin commented:
merged! can now move all the geometry-specific calculations to pre-processor directives rather than uniforms.

chrishavlin commented:
pre-processor directives now working

chrishavlin commented on yt_idv/shaders/known_uniforms.inc.glsl:
```diff
@@ -57,9 +57,19 @@ uniform vec4 curve_rgba;
 // isocontour control
 uniform int iso_num_layers;
 uniform float iso_layers[32];
+uniform bool iso_log;
+uniform float iso_min;
+uniform float iso_max;
```
hmm, not sure why these are here. probably should delete...

chrishavlin commented on yt_idv/shaders/ray_tracing.frag.glsl:
```diff
@@ -5,6 +5,8 @@ flat in vec3 right_edge;
 flat in mat4 inverse_proj;
 flat in mat4 inverse_mvm;
 flat in mat4 inverse_pmvm;
+flat in vec3 left_edge_cart;
```
missed a pre-processor check here

chrishavlin commented:
still do do: 

- [ ] docs docs docs
- [ ] another test or two would be nice -- maybe a series that checks different bounds.
- [ ] an example with a real dataset

chrishavlin commented on yt_idv/tests/test_yt_idv.py:
```diff
@@ -38,18 +35,6 @@ def osmesa_empty():
 
 
-@pytest.fixture()
-def image_store(request, extras, tmpdir):
```
note: this is now in conftest.py

chrishavlin commented on yt_idv/tests/conftest.py:
```diff
@@ -1,7 +1,24 @@
+
+
+@pytest.fixture()
+def image_store(request, extras, tmpdir):
```
copied from test_yt_idv.py so it can be used in other test files

chrishavlin commented:
Latest push includes some new logic to limit the options that show up in the GUI so that unsupported shaders and scene components are not shown. I also bumped the maximum sample factor when the data is in spherical coordinates.

chrishavlin commented:
Working on docs now...

chrishavlin commented:
Just adding an example of a dataset with refined grids. (code over [here](https://github.com/data-exp-lab/spherical_volume_rendering/blob/main/scratch/sample_data_creation/refined_grid_recursive.py))

https://github.com/user-attachments/assets/77772e94-f792-4106-8e7b-db9e622339f9

chrishavlin commented:
latest push adds block outlines via the CurveRendering data object
![snap_0000](https://github.com/user-attachments/assets/a55f415e-7904-449c-9189-ea4e150fbe6e)
![snap_0001](https://github.com/user-attachments/assets/38760e1c-c031-4b84-9b33-382921417e41)

chrishavlin commented:
added grid outlines now

https://github.com/user-attachments/assets/04004263-a743-43a3-a49a-8b2f97a9857d

chrishavlin commented:
Assuming tests still pass, I think this is ready!

I just updated the docs. and also added an additional pre-processor directive, `NONCARTESIAN_GEOM` that is used for directive checks in the rendering pipeline that would be common to any non-cartesian coordinate system. So the existing `SPHERICAL_GEOM` one is now only used where the coordiante conversion is defined in the fragment shader. This makes it easier to add support for other coordinate systems.

matthewturk commented on yt_idv/coordinate_utilities.pyx:
```diff
@@ -0,0 +1,526 @@
+            error code: 0 for success, 1 if something went wrong...
+
+        Note: this function is recursive! If either angular width is above a cutoff (of
+        0.2 radians or about 11 degrees), the element will be recursively divided.
```
Why 0.2 rad / 11 deg?

chrishavlin replied:
I... don't remember. Best guess looking at it now is that when playing with what to use, 10 degrees worked well without adding too much extra computation, but I wanted it to be a nice number in radians?

matthewturk commented on yt_idv/scene_components/blocks.py:
```diff
@@ -34,34 +34,58 @@ class BlockRendering(SceneComponent):
+                scene.data_objects.append(cc)
+                scene.components.append(cc_render)
+
+        # the following only work for cartesian data at present
```
Is this still true? :smile:

chrishavlin replied:
oh! no! the last few commits made these work :)

matthewturk commented on yt_idv/scene_components/blocks.py:
```diff
@@ -139,6 +163,18 @@ def draw(self, scene, program):
 
     def _set_uniforms(self, scene, shader_program):
+        if self.data._yt_geom_str == "spherical":
+            axis_id = self.data.data_source.ds.coordinates.axis_id
```
Great catch -- this is the kind of thing I usually miss doing.

matthewturk commented on yt_idv/shaders/ray_tracing.frag.glsl:
```diff
@@ -46,8 +80,18 @@ void main()
-    vec3 tr = (right_edge - camera_pos)*idir;
+    vec3 tl, tr, dx_cart;
+    #ifdef NONCARTESIAN_GEOM
+    tl = (left_edge_cart - camera_pos)*idir;
```
I'm not 100% sure but I *think* that this may be the source of the (mild!) artifacts we see when we viz `("index", "ones")` with projective integration.  I attached just-off-axis and almost-on-axis to demonstrate.

To be clear, these artifacts are not deal breakers!  But I'm wondering if this is a source of the artifacts, and if we could use more precise values to compute them.

![spherical-vr-off-edge](https://github.com/user-attachments/assets/be0af722-7604-47bc-8b9c-bbf12db94d12)
![spherical-vr-on-axis](https://github.com/user-attachments/assets/8410adf3-e5ee-46d4-90b3-07f5a9bc3140)

chrishavlin replied:
it could be! and i've banged my head on many things looking for the source of those artifacts but it didn't occur to me that it might a precision issue.

chrishavlin replied:
> I'm not 100% sure but I think that this may be the source of the (mild!) artifacts we see when we viz

You were 99% right! it was `dx_cart` (a few lines after this), see https://github.com/yt-project/yt_idv/pull/159#issuecomment-2839260924

matthewturk reviewed (approved):
This is amazing!  I've left a few comments, including one about possible artifacts.  But I think it's generally good to go!

chrishavlin commented on examples/spherical_amr_rendering.py:
```diff
@@ -0,0 +1,127 @@
+import argparse
```
oh! ya, that would make sense to change. will do that.

chrishavlin replied:
renamed the file

chrishavlin commented on yt_idv/scene_components/blocks.py:
```diff
@@ -34,34 +34,58 @@ class BlockRendering(SceneComponent):
+                scene.data_objects.append(cc)
+                scene.components.append(cc_render)
+
+        # the following only work for cartesian data at present
```
```suggestion
```

chrishavlin commented on yt_idv/scene_components/blocks.py:
```diff
@@ -139,6 +163,18 @@ def draw(self, scene, program):
+                "cart_bbox_max_width", self.data.cart_bbox_max_width
+            )
+            shader_program._set_uniform(
+                "cart_bbox_le", self.data.cart_bbox_le.astype("f4")
```
yes, the whole domain. i'll add a comment.

chrishavlin commented on yt_idv/coordinate_utilities.pxd:
```diff
@@ -0,0 +1,57 @@
+
+
+cdef class SphericalMixedCoordBBox(MixedCoordBBox):
+    cdef int get_cartesian_bbox(
```
you were correct, do not need to re-declare the super class signature.

chrishavlin commented:
@matthewturk I've been looking at that artifact issue on and off. and I'm not positive but I think it might be a limitation of the current implementation that uses the cartesian bounding box of each block as the ray start/end: if you imagine discretizing a ray, then the number of points along the ray that fall within a sphere will depend on the curvature of the sphere and the stepsize along the ray and there are situations that can lead to artifacts. I played around with the idea in 2D over in [this notebook](https://github.com/chrishavlin/yt_scratch/blob/master/notebooks/yt_idv_nbs/spherical_vol_render_integration_issue.ipynb), here's a figure that plots the integrated distance along a 2D ray that falls within a circle for rays that are parallel to the y axis along with the derivative of that distance with respect to the x position of the ray:

<img width="638" alt="Screenshot 2025-03-06 at 11 36 51 AM" src="https://github.com/user-attachments/assets/0ffcaa47-1ef7-416b-b6e8-6ecef0654afa" />


I think those little bumps may be the artifact we're seeing: there are spots where the integrated distance increases depending on how the circle's radius lines up with the discretized ray. Not sure there's a fix for this beyond working out the full intersection calculations.

matthewturk commented:
Let's imagine we *did* do the full intersection calculations.  Wouldn't we be able to do this only for the regions that actually intersect a ray -- i.e., only in the fragment shader, and so potentially with a reduced number of calculations?  The bounding boxes are how we compute which fragments to compute, but potentially couldn't we inside the fragment shader also use that to `discard` fragments, or compute the actual intersections?

Also, if there are computations (potentially, for instance, inverse trig functions, square roots, etc) that we can bundle we could do them inside a compute shader step, or otherwise pre-compute and store them.

chrishavlin commented:
> Let's imagine we did do the full intersection calculations. Wouldn't we be able to do this only for the regions that actually intersect a ray -- i.e., only in the fragment shader, and so potentially with a reduced number of calculations? The bounding boxes are how we compute which fragments to compute, but potentially couldn't we inside the fragment shader also use that to discard fragments, or compute the actual intersections?

yup! and that's what i've done in my attempts -- use the cartesian box check to initially discard fragments before doing the full intersection calculations. let me dig up my last attempt which I think was almost working and remind myself where things were going wrong...

chrishavlin commented on yt_idv/shaders/ray_tracing.frag.glsl:
```diff
@@ -46,8 +80,18 @@ void main()
+    #ifdef NONCARTESIAN_GEOM
+    tl = (left_edge_cart - camera_pos)*idir;
+    tr = (right_edge_cart - camera_pos)*idir;
+    dx_cart = right_edge_cart - left_edge_cart;
```
This line is wrong!!!!! in the cartesian case, `dx` is grid spacing in x, y, z. Here, it should be some approximation of cartesian distance across the spherical grid (maybe just `dr`, and arc lengths along the angular distances). haven't tested if it corrects the artifacts but it's definitely a reason for why the sample factor has to be bumped so high right now.

chrishavlin commented:
ok, just took care of some redundant re-scaling of the domain that I don't think is necessary. also had to replace a deprecated usage of `numpy.math` in the new cython code here. 

I think... this is good to go?? maybe I'll wait a day before merging...

chrishavlin commented on docs/coordinate_systems.rst:
```diff
@@ -0,0 +1,99 @@
+cartesian coordinates of the ray position are converted to the native coordinates of the data,
+which is used to sample the texture maps (which are stored in native coordinates). The initial
+implementaiton of the algorithm uses the ray entry/exit points of the cartesian bounding boxes,
+and so not all points along the ray are gaurenteed to lie within the bounds of the data -- though
```
```suggestion
and so not all points along the ray are gauranteed to lie within the bounds of the data -- though
```

chrishavlin commented on docs/coordinate_systems.rst:
```diff
@@ -0,0 +1,99 @@
+implementaiton of the algorithm uses the ray entry/exit points of the cartesian bounding boxes,
+and so not all points along the ray are gaurenteed to lie within the bounds of the data -- though
+these data points will be discarded during ray tracing, it does mean that a larger number of samples
+along the ray is required to gaurantee the underlying non-cartesian element is properly sampled compared
```
```suggestion
along the ray may be required to ensure the underlying non-cartesian element is properly sampled compared
```

chrishavlin commented on docs/coordinate_systems.rst:
```diff
@@ -0,0 +1,99 @@
+Add support to the rendering pipeline
+*************************************
+
+A number of changes are requirer related to the rendering pipeline for the ``block_rendering``
```
```suggestion
A number of changes are required related to the rendering pipeline for the ``block_rendering``
```

chrishavlin commented on docs/coordinate_systems.rst:
```diff
@@ -0,0 +1,99 @@
+passing along the additional vertex attributes related to the cartesian bounding boxes.
+
+Functionality specific to a coordinate system should be defined with a new pre-processor
+directive flag. For spherical coordiantes, this is ``SPHERICAL_GEOM``, and is used in
```
```suggestion
directive flag. For spherical coordinates, this is ``SPHERICAL_GEOM``, and is used in
```

chrishavlin reviewed (commented):
reading through the new docs one last time, found some typos

chrishavlin commented:
hmm. had a shader compilation failure just now with this code locally, not sure why the tests didnt catch it.

chrishavlin commented:
(so dont merge yet)

chrishavlin commented:
oh i see. the test images are incorrect, it's just that tests dont fail on compilation errors. that might be nice to change.
